File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.java
class Solution {
    public int[] solution(int rows, int columns, int[][] queries) {
        int[] answer = new int[queries.length];
        int[][] squre = new int[rows][columns];
        int temp = 1;
        int s = 0;
        int a = 0;
        int min = rows*columns;
        
        for(int i = 0;i<rows;i++){
            for(int j = 0;j<columns;j++){
                squre[i][j]=temp++;
            }
        }
        
        if(queries.length==1){
            answer[0]=squre[queries[0][0]-1][queries[0][1]-1];
            return answer;
        }
        
        for(int i = 0;i<queries.length;i++){
            s=queries[i][0]-1;
            temp=squre[s+1][queries[i][1]-1];
            for(int j = queries[i][1]-1;j<queries[i][3]-1;j++){
                if(squre[s][j]<min) min=squre[s][j];
                a=squre[s][j];
                squre[s][j]=temp;
                temp=a;
            }
            
            s=queries[i][3]-1;
            for(int j = queries[i][0]-1;j<queries[i][2]-1;j++){
                if(squre[j][s]<min) min=squre[j][s];
                a=squre[j][s];
                squre[j][s]=temp;
                temp=a;
            }
            
            s=queries[i][2]-1;
            for(int j = queries[i][3]-1;j>=queries[i][1];j--){
                if(squre[s][j]<min) min=squre[s][j];
                a=squre[s][j];
                squre[s][j]=temp;
                temp=a;
            }
            
            s=queries[i][1]-1;
            for(int j = queries[i][2]-1;j>=queries[i][0];j--){
                if(squre[j][s]<min) min=squre[j][s];
                a=squre[j][s];
                squre[j][s]=temp;
                temp=a;
            }
            
            answer[i]=min;
            min=rows*columns;
        }
        
        return answer;
    }
}

File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int DoQuery(vector<vector<int>> &matrix, const vector<int> &query)
{
    queue<pair<int, int>> q;
    for(int i = query[1]; i <= query[3]; ++i)
    {
        q.push({query[0], i});
    }

    for(int i = query[0] + 1; i <= query[2]; ++i)
    {
        q.push({i, query[3]});
    }

    for(int i = query[3] - 1; i >= query[1]; --i)
    {
        q.push({query[2], i});
    }

    for(int i = query[2] - 1; i >= query[0]; --i)
    {
        q.push({i, query[1]});
    }

    pair<int, int> front = q.front();
    int memory = matrix[front.first][front.second];
    int min = memory;
    int num;
    while(!q.empty())
    {
        q.pop();
        if(q.empty()) break;
        front = q.front();
        num = matrix[front.first][front.second];
        if(num == 0) break;
        if(num < min) min = num;
        matrix[front.first][front.second] = memory;
        memory = num;
    }
    return min;
}

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;
    vector<vector<int>> matrix(rows + 1, vector<int>(columns + 1));
    for(int i = 1; i <= rows; ++i)
    {
        for(int j = 1; j <= columns; ++j)
            matrix[i][j] = ((i - 1) * columns + j);
    }
    for(const auto query : queries)
    {
        answer.emplace_back(DoQuery(matrix, query));
    }
    return answer;
}